src/analysis/special_functions/bernstein.lean

```diff
@@ -0,0 +1,377 @@
+  have h' : (n : ℝ) ≠ 0 := (ne_of_lt h).symm,
+  apply_fun (λ x : ℝ, x * n) using group_with_zero.mul_right_injective h',
+  apply_fun (λ x : ℝ, x * n) using group_with_zero.mul_right_injective h',
+  dsimp,
+  conv_lhs { simp only [finset.sum_mul, z], },
+  conv_rhs { rw div_mul_cancel _ h', },
+  have := bernstein_polynomial.variance ℝ n,
+  apply_fun (λ p, polynomial.aeval (x : ℝ) p) at this,
+  simp [alg_hom.map_sum, finset.sum_range, polynomial.nat_smul] at this,
+  convert this using 1,
+  { congr' 1, funext k,
+    rw [mul_comm _ (n : ℝ), mul_comm _ (n : ℝ), ←mul_assoc, ←mul_assoc],
+    congr' 1,
+    field_simp [h],
+    ring, },
+  { ring, },
+end
+
+end bernstein
+
+open bernstein
+
+local postfix `/ₙ`:2000 := z
+
+/--
+The `n`-th approximation of a continuous function on `[0,1]` by Bernstein polynomials,
+given by `∑ k, f (k/n) * bernstein n k x`.
+-/
+def bernstein_approximation (n : ℕ) (f : I →ᵇ ℝ) : I →ᵇ ℝ :=
+∑ k : fin (n+1), f k/ₙ • bernstein n k
+
+namespace bernstein_approximation
+
+@[simp] lemma apply (n : ℕ) (f : I →ᵇ ℝ) (x : I) :
+  bernstein_approximation n f x = ∑ k : fin (n+1), f k/ₙ * bernstein n k x :=
+by simp [bernstein_approximation]
+
+/--
+The set of points `k` so `k/n` is within `δ` of `x`.
+-/
+def S (f : I →ᵇ ℝ) (ε : ℝ) (h : 0 < ε) (n : ℕ) (x : I) : finset (fin (n+1)) :=
+{ k : fin (n+1) | dist k/ₙ x < f.modulus (ε/2) (half_pos h) }.to_finset
+
+/--
+If `k ∈ S`, then `f(k/n)` is close to `f x`.
+-/
+lemma lt_of_mem_S
+  {f : I →ᵇ ℝ} {ε : ℝ} {h : 0 < ε} {n : ℕ} {x : I} {k : fin (n+1)} (m : k ∈ S f ε h n x) :
+  abs (f k/ₙ - f x) < ε/2 :=
+begin
+  apply f.dist_lt_of_dist_lt_modulus (ε/2) (half_pos h),
+  simpa [S] using m,
+end
+
+/--
+If `k ∉ S`, then as `δ ≤ abs (x - k/n)`, we have the inequality `1 ≤ δ^-2 * (x - k/n)^2`.
+This particular formulation will be helpful later.
+-/
+lemma le_of_mem_S_compl
+  {f : I →ᵇ ℝ} {ε : ℝ} {h : 0 < ε} {n : ℕ} {x : I} {k : fin (n+1)} (m : k ∈ (S f ε h n x)ᶜ) :
+  (1 : ℝ) ≤ (f.modulus (ε/2) (half_pos h))^(-2 : ℤ) * (x - k/ₙ) ^ 2 :=
+begin
+  simp only [finset.mem_compl, not_lt, set.mem_to_finset, set.mem_set_of_eq, S] at m,
+  field_simp,
+  erw [le_div_iff (pow_pos f.modulus_pos 2), one_mul],
+  apply sqr_le_sqr,
+  rw abs_eq_self.mpr (le_of_lt f.modulus_pos),
+  rw [dist_comm] at m,
+  exact m,
+end
+
+end bernstein_approximation
+
+open bernstein_approximation
+open bounded_continuous_function
+open filter
+
+open_locale topological_space
+
+/--
+This is the proof given in [Richard Beals' *Analysis, an introduction*][beals-analysis], §7D,
+and reproduced on wikipedia.
+-/
+theorem bernstein_approximation_uniform (f : I →ᵇ ℝ) :
+  tendsto (λ n : ℕ, bernstein_approximation n f) at_top (𝓝 f) :=
+begin
+  apply normed_group.tendsto_at_top.mpr,
+  intros ε h,
+  let δ := f.modulus (ε/2) (half_pos h),
+  let N : ℕ := _, use N, -- We postpone choosing `n` until we've obtained an explicit estimate.
+  intros n nh,
+  have npos : 0 < (n : ℝ) := by exact_mod_cast (pos_of_gt nh),
```

Comment:
```suggestion
  simp only [metric.nhds_basis_ball.tendsto_right_iff, metric.mem_ball, dist_eq_norm],
  intros ε h,
  let δ := f.modulus (ε/2) (half_pos h),
  have nhds_zero := tendsto_const_div_at_top_nhds_0_nat (2 * ∥f∥ * δ ^ (-2 : ℤ)),
  filter_upwards [nhds_zero.eventually (gt_mem_nhds (half_pos h)), eventually_gt_at_top 0],
  intros n nh npos',
  have npos : 0 < (n:ℝ) := by exact_mod_cast npos',
```
This is the "filter-y" way --you end up with a hypothesis `nh`, ready to go for the last step.  Do you hate it? :)

Reply:
No it's great. The only problem is that I never would have been able to write line 275. (Or 278 for that matter!)